### PR TITLE
ANSI rendering of rst markup.

### PR DESCRIPTION
**Pull request by anntzer**
See #22. Note that this PR depends on #20 (the first few commits are from that PR).

### CODE REVIEW

**Comment by anntzer** [comment]
Rebased.

File: setup.cfg [hunk 1]
```diff
@@ -5,4 +5,4 @@ universal=1
 source=defopt
 
 [coverage:report]
-fail_under=98
+fail_under=97
```
Comment (evanunderscore) [inline setup.cfg]: Why did you need to change this? I don't think your changes reduced coverage.

File: setup.py [hunk 2]
```diff
@@ -19,12 +19,13 @@
     install_requires=[
         'docutils',
         'sphinxcontrib-napoleon>=0.5.1',
-        'funcsigs;python_version<"3"',
+        'funcsigs;python_version<"3.3"',
```
Comment (evanunderscore) [inline setup.py]: For the record, I don't intend to support 3.0-3.2, but I don't have a problem with changing this.

File: defopt.py [hunk 3]
```diff
@@ -312,8 +319,28 @@ def _parse_doc(func):
     return _Doc(doctext, tuples)
 
 
+def _itertext_ansi(node):
+    # Implementation modified from `ElementTree.Element.itertext`.
+    tag = node.tag
+    if not isinstance(tag, str) and tag is not None:
```
Comment (evanunderscore) [inline defopt.py]: This is `basestring` in the Python 2 version of `itertext`. We should account for that.

File: defopt.py [hunk 4]
```diff
@@ -25,6 +25,13 @@
 except ImportError:  # pragma: no cover
     from funcsigs import signature as _inspect_signature
 
+try:
+    import colorama
+except ImportError:  # pragma: no cover
+    pass
+else:
+    colorama.init()  # pragma: no cover
```
Comment (evanunderscore) [inline defopt.py]: I know this is the way colorama tells you to use it, but since defopt promises to not modify the functions given, I think it's also reasonable to expect it to have no side effects (especially when you're not even using the CLI). Instead, I think we should use `colorama.colorama_text` to wrap just the call to `parser.parse_args` inside `run`.

File: defopt.py [hunk 5]
```diff
@@ -312,8 +319,28 @@ def _parse_doc(func):
     return _Doc(doctext, tuples)
 
 
+def _itertext_ansi(node):
+    # Implementation modified from `ElementTree.Element.itertext`.
+    tag = node.tag
+    if not isinstance(tag, str) and tag is not None:
+        return
+    ansi_code = {"emphasis": "\033[1m", "strong": "\033[3m"}.get(tag, "")
```
Comment (evanunderscore) [inline defopt.py]: You have these reversed - 1 is bold, 3 is italic.

**Comment by evanunderscore** [comment]
Thanks for your work. Overall the implementation is great, there are just a few things I'd like to get cleaned up before I merge this in.

File: setup.cfg (hunk 1 above)
Comment (anntzer) [inline setup.cfg]: The `return` in
```
    if not isinstance(tag, str) and tag is not None:
        return
```
doesn't get exercised (and I don't think it's worth writing a test to specifically test it -- it comes straight from the upstream implementation).  That's enough for dropping the coverage.

File: setup.py (hunk 2 above)
Comment (anntzer) [inline setup.py]: I don't plan to use them either, but that's just a more convenient way to say "function signatures were introduced in 3.3"

File: defopt.py (hunk 3 above)
Comment (anntzer) [inline defopt.py]: fixed

File: defopt.py (hunk 4 above)
Comment (anntzer) [inline defopt.py]: fixed

File: defopt.py (hunk 5 above)
Comment (anntzer) [inline defopt.py]: fixed

File: defopt.py [hunk 6]
```diff
@@ -312,8 +321,29 @@ def _parse_doc(func):
     return _Doc(doctext, tuples)
 
 
+def _itertext_ansi(node):
+    # Implementation modified from `ElementTree.Element.itertext`.
+    tag = node.tag
+    if (not isinstance(tag, (basestring if sys.version_info < (3,) else str))
```
Comment (evanunderscore) [inline defopt.py]: Sorry to keep getting you to change this, but this is a bit cluttered for my taste. Can you instead add a definition for basestring at the top with the rest of the compatibility stuff?

File: defopt.py [hunk 7]
```diff
@@ -12,6 +12,7 @@
 from argparse import (SUPPRESS, ArgumentParser, RawTextHelpFormatter,
                       ArgumentDefaultsHelpFormatter)
 from collections import defaultdict, namedtuple, OrderedDict
+import contextlib
```
Comment (evanunderscore) [inline defopt.py]: These are split into separate subgroups for `import x` and `from x import y`. I was aiming for what you get from isort, though I'm not actually running the tool so this may differ slightly.

**Comment by anntzer** [comment]
done

**Comment by evanunderscore** [comment]
Perfect, thanks!
